libglusterfs/src/mem-pool.c

```diff
@@ -578,58 +576,40 @@ mem_pools_postfini(void)
      *       so the memory will be released anyway by the system. */
 }
 
-/* Call mem_pools_init() once threading has been configured completely. This
- * prevent the pool_sweeper thread from getting killed once the main() thread
- * exits during deamonizing. */
+static void
+mem_pools_init_once(void)
+{
+    if (gf_thread_create(&sweeper_tid, NULL, pool_sweeper, NULL, "memsweep"))
+        GF_ABORT("can't create memsweep thread");
+}
+
 void
 mem_pools_init(void)
 {
-    pthread_mutex_lock(&init_mutex);
-    if ((init_count++) == 0) {
-        (void)gf_thread_create(&sweeper_tid, NULL, pool_sweeper, NULL,
-                               "memsweep");
+    static pthread_once_t init_once = PTHREAD_ONCE_INIT;
+    pthread_once(&init_once, mem_pools_init_once);
+}
 
-        init_done = GF_MEMPOOL_INIT_LATE;
-    }
-    pthread_mutex_unlock(&init_mutex);
+static void
+mem_pools_fini_once(void)
+{
+    pthread_mutex_lock(&sweeper_mutex);
+    pthread_cond_signal(&sweeper_cond);
```

Comment:
If the sweeper thread is not blocked on the condition variable at this point, the signal will be missed and the thread won't be stopped.

Reply:
OK switch to self-pipe instead.